From 2cf84d954eb5312ac49b058adc05445c715d09b6 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Thu, 25 Sep 2014 13:23:16 -0400
Subject: [acpi] container: Fix error code path in container_device_attach()

Message-id: <1411651397-12618-1-git-send-email-prarit@redhat.com>
Patchwork-id: 96720
O-Subject: [RHEL7.1 PATCH BZ 1128634 3/2] ACPI / container: Fix error code path in container_device_attach()
Bugzilla: 1128634
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1128634

Patch suggested as part of initial review by dnelson.

commit 0f6aa09e40c38d734f5d70762056116ca4a76126
Author: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
Date:   Wed Feb 12 00:52:49 2014 +0100

    ACPI / container: Fix error code path in container_device_attach()

    To avoid leaking memory on errors from device_register(), do a
    put_device() on the device object in question in the error code
    path of container_device_attach().

    Fixes: caa73ea158de (ACPI / hotplug / driver core: Handle containers in a special way)
    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
    Reviewed-by: Yasuaki Ishimatsu <isimatu.yasuaki@jp.fujitsu.com>

Cc: Lenny Szubowicz <lszubowi@redhat.com>
Cc: Stefan Assmann <sassmann@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/acpi/container.c b/drivers/acpi/container.c
index be9a3b8..3ed6fd0 100644
--- a/drivers/acpi/container.c
+++ b/drivers/acpi/container.c
@@ -80,9 +80,10 @@ static int container_device_attach(struct acpi_device *adev,
  ACPI_COMPANION_SET(dev, adev);
  dev->release = acpi_container_release;
  ret = device_register(dev);
- if (ret)
+ if (ret) {
+  put_device(dev);
   return ret;
-
+ }
  adev->driver_data = dev;
  return 1;
 }
-- 
1.7.1